### ISSUE TITLE
[libpq] build failure

### ISSUE THREAD

**Issue by mkaring** [microsoft/vcpkg#18527]
**Host Environment**
- OS: Windows
- Compiler: Visual Studio 2019 16.10.2
- Port files: 6e024e744e7717c06ddacd5089401109c6298553

**To Reproduce**
`.\vcpkg\vcpkg.exe install libpq:x64-windows`

**Failure logs**
```
Computing installation plan...
The following packages will be built and installed:
    libpq[core,openssl,zlib]:x64-windows -> 12.2#16
Detecting compiler hash for triplet x64-windows...
Could not locate cached archive: C:\Users\m.karing\AppData\Local\vcpkg\archives\aa\aa6fb6fc8b1c8d3e2fdbfc43d2931de3974b9067.zip
Starting package 1/1: libpq:x64-windows
Building package libpq[core,openssl,zlib]:x64-windows...
-- Using cached C:/build/vcpkg/downloads/postgresql-12.2.tar.bz2
-- Cleaning sources at C:/build/vcpkg/buildtrees/libpq/src/tgresql-12-c160d9fcbb.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/build/vcpkg/downloads/postgresql-12.2.tar.bz2
-- Applying patch patches/windows/install.patch
-- Applying patch patches/windows/win_bison_flex.patch
-- Applying patch patches/windows/openssl_exe_path.patch
-- Applying patch patches/windows/Solution.patch
-- Applying patch patches/windows/MSBuildProject_fix_gendef_perl.patch
-- Applying patch patches/windows/msgfmt.patch
-- Applying patch patches/windows/python_lib.patch
-- Applying patch patches/windows/fix-compile-flag-Zi.patch
-- Applying patch patches/windows/minimize_install.patch
-- Using source at C:/build/vcpkg/buildtrees/libpq/src/tgresql-12-c160d9fcbb.clean
-- Copying libpq source files to C:/build/vcpkg/buildtrees/libpq/x64-windows-rel...
-- Copying libpq source files... done
CMake Deprecation Warning at scripts/ports.cmake:16 (message):
  vcpkg_apply_patches has been deprecated in favor of the `PATCHES` argument
  to `vcpkg_from_*`.
Call Stack (most recent call first):
  scripts/cmake/vcpkg_apply_patches.cmake:18 (z_vcpkg_deprecation_message)
  ports/libpq/portfile.cmake:127 (vcpkg_apply_patches)
  scripts/ports.cmake:142 (include)


-- Applying patch patches/windows/Solution_RELEASE.patch
-- Applying patch patches/windows/python3_build_RELEASE.patch
-- Patches applied!
-- Building libpq x64-windows-RELEASE...
-- Building libpq x64-windows-RELEASE... done
-- Building libecpg_compat x64-windows-RELEASE...
-- Building libecpg_compat x64-windows-RELEASE... done
-- Installing libpq x64-windows-RELEASE...
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:105 (message):
    Command failed: C:/build/vcpkg/downloads/tools/perl/5.30.0.1/perl/bin/perl.exe install.pl C:/build/vcpkg/packages/libpq_x64-windows client
    Working Directory: C:/build/vcpkg/buildtrees/libpq/x64-windows-rel/src/tools/msvc
    Error code: 2
    See logs for more information:
      C:\build\vcpkg\buildtrees\libpq\install-x64-windows-RELEASE-err.log

Call Stack (most recent call first):
  ports/libpq/portfile.cmake:233 (vcpkg_execute_required_process)
  scripts/ports.cmake:142 (include)


Error: Building package libpq:x64-windows failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: libpq:x64-windows
  Vcpkg version: 2021-05-05-9f849c4c43e50d1b16186ae76681c27b0c1be9d9
```
[install-x64-windows-RELEASE-err.log](https://github.com/microsoft/vcpkg/files/6675755/install-x64-windows-RELEASE-err.log)

**Comment by PhoebeHui** [comment]
```
Could not find debug or release binaries at C:\build\vcpkg\buildtrees\libpq\x64-windows-rel\src\tools\msvc/Install.pm line 77.
```

@mkaring, I can't repro this issue locally, it built successfully for me. 

looking at src\tools\msvc/Install.pm line 68 ~ line 77, it looks something wrong here. Could you also attach build-libpq-x64-windows-RELEASE-out.log? 
```
	my $conf = "";
	if (-d "debug")
	{
		$conf = "debug";
	}
	if (-d "release")
	{
		$conf = "release";
	}
	die "Could not find debug or release binaries" if ($conf eq "");

```

**Comment by mkaring** [comment]
I cleaned the entire storage structure of vcpkg and installed libpq again. This time it worked. I'm not sure why. Maybe a side effect of another package or of the actual folder structure.
I give it another shot trying to reproduce it. I think the reason for the issue was, that the directory structure that is resolved in the line(s) above the block you copied did not line up correctly.

I'll try to reproduce it by the end of the week. If I can't, I'll close the issue and assume it magically fixed itself.

**Comment by mkaring** [comment]
I can't reproduce the issue anymore.

**Comment by mkaring** [comment]
Okay. I found the cause for the issue, I'm able to reproduce and explain is.
The problem is caused because the root directory the `vcpkg` repository is checked out into contains a `configure` file.

Now the relevant file system structure looks like this:
```
C:\build\
C:\build\configure    <-- important
C:\build\vcpkg\
C:\build\vcpkg\buildtrees\
C:\build\vcpkg\buildtrees\libpq\
C:\build\vcpkg\buildtrees\libpq\x64-windows-rel\
C:\build\vcpkg\buildtrees\libpq\x64-windows-rel\configure
C:\build\vcpkg\buildtrees\libpq\x64-windows-rel\Release\ <-- looking for this one
C:\build\vcpkg\buildtrees\libpq\x64-windows-rel\src\
C:\build\vcpkg\buildtrees\libpq\x64-windows-rel\src\tools\
C:\build\vcpkg\buildtrees\libpq\x64-windows-rel\src\tools\msvc\   <-- working dir
C:\build\vcpkg\buildtrees\libpq\x64-windows-rel\src\tools\msvc\Install.pm
```

The working directory at the very beginning is:
```
C:\build\vcpkg\buildtrees\libpq\x64-windows-rel\src\tools\msvc\
```

Now first the following perl line is executed:
```perl
chdir("../../..")    if (-f "../../../configure");
```
Three directories up is
```
C:\build\vcpkg\buildtrees\libpq\x64-windows-rel\
```
This directory contains a `configure` file, so the directory is changed.
Now the next line is:
```perl
chdir("../../../..") if (-f "../../../../configure");
```
Four directories up from where we are is
```
C:\build\
```
This directory contains a `configure` file as well. Meaning this line is executed as well. So now the working directory is far away from where it needs to be and so the next lines, looking for a `Release` or `Debug` directory fail.

**Comment by PhoebeHui** [comment]
@mkaring, thanks for your detail repro steps, I could repro this issue locally now.

**Comment by longhuan2018** [comment]
The bug has been fixed in version 13.2.
Install.pm file of libpq version 12.2 
```
	chdir("../../..")    if (-f "../../../configure");
	chdir("../../../..") if (-f "../../../../configure");
```

Install.pm file of libpq version 13.2
```
	# Move to the root path depending on the current location.
	if (-f "../../../configure")
	{
		chdir("../../..");
	}
	elsif (-f "../../../../configure")
	{
		chdir("../../../..");
	}
```
So we can fix this problem by adding the following contents in ports/libpq/patches/windows/install.patch
```
@@ -63,8 +63,16 @@ sub Install
 		do "./config.pl" if (-f "config.pl");
 	}
 
-	chdir("../../..")    if (-f "../../../configure");
-	chdir("../../../..") if (-f "../../../../configure");
+	# Move to the root path depending on the current location.
+	if (-f "../../../configure")
+	{
+		chdir("../../..");
+	}
+	elsif (-f "../../../../configure")
+	{
+		chdir("../../../..");
+	}
+
 	my $conf = "";
 	if (-d "debug")
 	{
```
